pkg/planner/core/rule_outer_to_inner_join.go

```diff
@@ -0,0 +1,185 @@
+		switchChild = true
+	}
+
+	// First, simplify this join
+	if p.JoinType == LeftOuterJoin || p.JoinType == RightOuterJoin {
+		canBeSimplified := false
+		for _, expr := range predicates {
+			isOk := isNullFiltered(p.SCtx(), innerTable.Schema(), expr, outerTable.Schema())
+			if isOk {
+				canBeSimplified = true
+				break
+			}
+		}
+		if canBeSimplified {
+			p.JoinType = InnerJoin
+		}
+	}
+
+	// Next simplify join children
+
+	combinedCond := mergeOnClausePredicates(p, predicates)
+	if p.JoinType == LeftOuterJoin || p.JoinType == RightOuterJoin {
+		innerTable = innerTable.ConvertOuterToInnerJoin(combinedCond)
+		outerTable = outerTable.ConvertOuterToInnerJoin(predicates)
+	} else if p.JoinType == InnerJoin || p.JoinType == SemiJoin {
+		innerTable = innerTable.ConvertOuterToInnerJoin(combinedCond)
+		outerTable = outerTable.ConvertOuterToInnerJoin(combinedCond)
+	} else if p.JoinType == AntiSemiJoin {
+		innerTable = innerTable.ConvertOuterToInnerJoin(predicates)
+		outerTable = outerTable.ConvertOuterToInnerJoin(combinedCond)
+	} else {
+		innerTable = innerTable.ConvertOuterToInnerJoin(predicates)
+		outerTable = outerTable.ConvertOuterToInnerJoin(predicates)
+	}
+
+	if switchChild {
+		p.SetChild(0, outerTable)
+		p.SetChild(1, innerTable)
+	} else {
+		p.SetChild(0, innerTable)
+		p.SetChild(1, outerTable)
+	}
+
+	return p
+}
+
+func (*convertOuterToInnerJoin) name() string {
+	return "convert_outer_to_inner_joins"
+}
+
+// ConvertOuterToInnerJoin implements base.LogicalPlan ConvertOuterToInnerJoin interface.
+func (s *LogicalSelection) ConvertOuterToInnerJoin(predicates []expression.Expression) base.LogicalPlan {
+	p := s.self.(*LogicalSelection)
+	combinedCond := append(predicates, p.Conditions...)
+	child := p.Children()[0]
+	child = child.ConvertOuterToInnerJoin(combinedCond)
+	p.SetChildren(child)
+	return p
+}
+
+// ConvertOuterToInnerJoin implements base.LogicalPlan ConvertOuterToInnerJoin interface.
+func (s *LogicalProjection) ConvertOuterToInnerJoin(predicates []expression.Expression) base.LogicalPlan {
+	p := s.self.(*LogicalProjection)
+	canBePushed, _ := BreakDownPredicates(p, predicates)
+	child := p.Children()[0]
+	child = child.ConvertOuterToInnerJoin(canBePushed)
+	p.SetChildren(child)
+	return p
+}
+
+// allConstants checks if only the expression has only constants.
+func allConstants(expr expression.Expression) bool {
+	switch v := expr.(type) {
+	case *expression.ScalarFunction:
+		for _, arg := range v.GetArgs() {
+			if !allConstants(arg) {
+				return false
+			}
+		}
+		return true
+	case *expression.Constant:
+		return true
+	}
+	return false
+}
+
+// isNullFiltered takes care of complex predicates like this:
+// isNullFiltered(A OR B) = isNullFiltered(A) AND isNullFiltered(B)
+// isNullFiltered(A AND B) = isNullFiltered(A) OR isNullFiltered(B)
+func isNullFiltered(ctx base.PlanContext, innerSchema *expression.Schema, predicate expression.Expression, outerSchema *expression.Schema) bool {
```

Comment:
Should we add this part logic into util.IsNullRejected() ?

Reply:
We could but it is a common routine to other parts of the code and could be risky to do so. May be a subsequent PR can do that.

Reply:
+1, better to put these 2 functions together